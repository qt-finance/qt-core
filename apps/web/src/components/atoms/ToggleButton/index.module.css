.button {
	background: var(--standard);
	padding: 4px 10px;
	border-radius: 5px;
	border: none;
	color: var(--secondary);

	&.reverse {
		background: var(--secondary);
		color: var(--standard);
	}
}
