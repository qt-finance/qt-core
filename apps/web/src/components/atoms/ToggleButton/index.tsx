import React from 'react';
import classnames from 'classnames';

import { useBoolean } from 'util/hook';

import style from './index.module.css';

interface ToggleButtonProperty {
	onClose: () => void;
	onOpen: () => void;
	closeTitle?: string;
	openTitle?: string;
}

const ToggleButton: React.FC<ToggleButtonProperty> = ({
	onClose,
	onOpen,
	closeTitle = 'Close',
	openTitle = 'Open',
}) => {
	const [isOpen, { toggle }] = useBoolean({
		onTrue: onOpen,
		onFalse: onClose,
		defaultBoolean: true,
	});

	return (
		<button
			type="button"
			className={classnames(style.button, {
				[style.reverse]: !isOpen,
			})}
			onClick={toggle}
		>
			{isOpen ? closeTitle : openTitle}
		</button>
	);
};

export default ToggleButton;
