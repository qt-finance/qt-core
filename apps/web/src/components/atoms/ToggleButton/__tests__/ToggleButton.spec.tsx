import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';

import ToggleButton from 'components/atoms/ToggleButton';

import * as stories from './ToggleButton.stories';

const testCases = Object.values(composeStories(stories)).map(Story => [
	// The ! is necessary in Typescript only, as the property is part of a partial type
	Story.storyName!, // eslint-disable-line @typescript-eslint/no-non-null-assertion
	Story,
]);
// Batch snapshot testing
test.each(testCases)('Renders %s story', async (_storyName, Story) => {
	const tree = await render(<Story />);
	expect(tree.baseElement).toMatchSnapshot();
});

test('should have default value: Close', () => {
	const { container } = render(<ToggleButton onOpen={() => {}} onClose={() => {}} />);
	expect((container.firstChild as HTMLButtonElement).innerHTML).toMatch('Close');
});

// Create custom snapshot testing
test('should become Open after click', () => {
	const { container } = render(<ToggleButton onOpen={() => {}} onClose={() => {}} />);

	expect(container).toMatchSnapshot();

	// manually trigger the callback
	fireEvent.click(container.firstChild as HTMLButtonElement);

	expect((container.firstChild as HTMLButtonElement).innerHTML).toMatch('Open');
	// re-rendering would become Open
	expect(container).toMatchSnapshot();

	// manually trigger the callback
	fireEvent.click(container.firstChild as HTMLButtonElement);

	expect((container.firstChild as HTMLButtonElement).innerHTML).toMatch('Close');
	// re-rendering would become Close
	expect(container).toMatchSnapshot();
});
