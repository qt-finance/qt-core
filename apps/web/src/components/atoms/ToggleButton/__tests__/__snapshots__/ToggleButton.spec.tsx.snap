// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Interactive story 1`] = `
<body>
  <div>
    <button
      class="button"
      type="button"
    >
      Close Title
    </button>
  </div>
</body>
`;

exports[`should become Open after click 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Close
  </button>
</div>
`;

exports[`should become Open after click 2`] = `
<div>
  <button
    class="button reverse"
    type="button"
  >
    Open
  </button>
</div>
`;

exports[`should become Open after click 3`] = `
<div>
  <button
    class="button"
    type="button"
  >
    Close
  </button>
</div>
`;
